mcstatus/protocol/connection.py

```diff
@@ -253,19 +256,20 @@ def __del__(self):
 
 
 class TCPAsyncSocketConnection(AsyncReadConnection):
-    reader = None
-    writer = None
-    timeout = None
+    # These will only be None until connect is called, ignore the None type assignment
+    reader: asyncio.StreamReader = None  # type: ignore[assignment]
+    writer: asyncio.StreamWriter = None  # type: ignore[assignment]
+    timeout: float = None  # type: ignore[assignment]
```

Comment:
`typing.Optional` exists for cases like this
```suggestion
    reader: Optional[asyncio.StreamReader] = None
    writer: Optional[asyncio.StreamWriter] = None
    timeout: Optional[float] = None
```

Reply:
I'm aware of that, however there was a reason I did this instead. I wanted to avoid further type errors down the line as if these types were defined as typing.Optional, every time you'd do for example `reader.x` the type checker would complain that `NoneType` has no attribute `x`.

The way around this while still passing type checkers would be to add these kinds of type ignore lines everywhere these variables will actually be used and ignore these, however I think that's a lot harder to maintain and it's just more unnecessary code clutter with these ignore comments. An alternative approach would be type narrowing by having a condition that checks if the type is none and raises an exception/returns prematurely (or having the logic in an exception that ensures the type isn't none). That would look like this:
```py
if reader is None:
    raise SomeException
``` 

Should I really still make this optional? If yes, which of these approaches should I take?

Reply:
pytype passed when I made this modification locally. Are you sure this would cause an issue for type checking?

Reply:
It is a typing issue, and both pyright and mypy do catch it as such, I'm not sure why pytype wouldn't be catching this but it's certainly a typing issue. The type checkers can't know that a function which sets these values is expected to run after initialization, if that's desired, it should be in init, not in a separate function. Because of that type-checkers can't narrow that type down and drop the default `None`, and obviously, accessing an attribute that only one of the union types have is a typing issue.

Minimal reproducable example:
```py
from typing import Union

x: Union[dict, list] = [1, 2, 3]

# This obviously can't work, `dict` doesn't support `append`
# it's the same error as None not having some func/attribute while the other type does
# it simply can't be right type-wise
x.append(5)
```

Reply:
`Optional[T]` is defined as `Union[T, type(None)]`, so what you are saying makes sense to me now
https://github.com/python/cpython/blob/7c5b01b5101923fc38274c491bd55239ee9f0416/Lib/typing.py#L363-L365